dags/data_retention_delete_dag.py

```diff
@@ -0,0 +1,86 @@
+"""
+Implements a retention policy by dropping expired partitions
+
+A detailed tutorial is available at https://community.crate.io/t/cratedb-and-apache-airflow-implementation-of-data-retention-policy/913
+
+Prerequisites
+-------------
+In CrateDB, tables for storing retention policies need to be created once manually.
+See the file setup/data_retention_schema.sql in this repository.
+"""
+import datetime
+import json
+import logging
+from pathlib import Path
+from airflow import DAG
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.providers.postgres.hooks.postgres import PostgresHook
+from airflow.operators.python import PythonOperator
+
+
+def get_policies(logical_date):
+    pg_hook = PostgresHook(postgres_conn_id="cratedb_connection")
+    sql = Path('include/data_retention_retrieve_delete_policies.sql') \
+              .read_text().format(date=logical_date)
+    records = pg_hook.get_records(sql=sql)
+
+    return json.dumps(records)
+
+
+def map_policy(policy):
+    return {
+        "table_fqn": policy[0],
+        "column_name": policy[1],
+        "partition_value": policy[2],
+    }
+
+
+def delete_partitions(ti):
+    retention_policies = ti.xcom_pull(task_ids="retrieve_retention_policies")
+    policies_obj = json.loads(retention_policies)
+
+    for policy in policies_obj:
+        partition = map_policy(policy)
+
+        logging.info("Deleting partition %s = %s for table %s",
+                     partition["column_name"],
+                     partition["partition_value"],
+                     partition["table_fqn"],
+        )
+
+        PostgresOperator(
+            task_id="delete_from_{table}_{partition}_{value}".format(
+                table=partition["table_fqn"],
+                partition=partition["column_name"],
+                value=partition["partition_value"],
+            ),
+            postgres_conn_id="cratedb_connection",
+            sql=Path('include/data_retention_delete.sql').read_text().format(
```

Comment:
This whole `Path` pattern that we used is a workaround. The `PostgresOperator` is actually supposed to accept a file name in the `sql` parameter directly and take care of loading it. But we faced an issue that the `PostgresOperator` was unable to load the file when it is nested in a `PythonOperator`. See also [this](https://stackoverflow.com/questions/67229594/airflow-running-postgresoperator-within-a-pythonoperator-with-taskflow-api) Stackoverflow post from another user. Hence the workaround using `Path`/`format`.
Potentially this could also be addressed/become irrelevant with AIP-42?

Reply:
Ah, understood. In that case, it seems like a very clever workaround! AIP-42 should address this so that you don't have to nest operators anymore; creating tasks from the results of a database query is one of the specific use-cases we're looking at solving.